lizmap/modules/lizmap/lib/Request/RemoteStorageRequest.php

```diff
@@ -0,0 +1,361 @@
+<?php
+/**
+ * Utilities for upload, delete and retreive resources from remote storage.
+ *
+ * @author
+ * @copyright 2012-2016 3liz
+ *
+ * @see      http://3liz.com
+ *
+ * @license Mozilla Public License : http://www.mozilla.org/MPL/
+ */
+
+namespace Lizmap\Request;
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Psr7;
+
+class RemoteStorageRequest
+{
+    protected static $webDAVNamespace = 'DAV:';
+
+    protected static $fileNameExpression = 'file_name(@selected_file_path)';
+    protected static $appContext;
+
+    /**
+     * @param string $storageUrl remote destination path
+     * @param string $file       the file path to upload
+     *
+     * @return array{0: null|string, 1: int, 2: string} Array url of the uploaded file (0: string), HTTP code (1: int), message (2: string)
+     */
+    public static function uploadToWebDAVStorage($storageUrl, $file)
+    {
+        $resource = null;
+        $http_code = 400;
+        $message = null;
+        $returnUrl = null;
+        // checks on file
+        if (!is_file($file)) {
+            \jLog::log($file.' The file is not valid', 'error');
+
+            return array(null, 400, 'The file is not valid');
+        }
+
+        try {
+            $resource = Psr7\Utils::tryFopen($file, 'r');
+        } catch (\RuntimeException $e) {
+            \jLog::log($e->getMessage(), 'error');
+
+            return array(null, 400, 'Error on file upload');
+        }
+
+        // get profile connection parameters
+        $profile = self::getProfile('webdav');
+
+        if ($profile) {
+            // baseUri check
+            if (strpos($storageUrl, $profile['baseUri']) === 0) {
+                $opt = array();
+                $headers = array();
+                $headers['User-Agent'] = 'lizmap-user-agent';
+
+                if (array_key_exists('user', $profile) && array_key_exists('password', $profile)) {
+                    $opt['auth'] = array();
+                    array_push($opt['auth'], $profile['user']);
+                    array_push($opt['auth'], $profile['password']);
+                }
+                $opt['headers'] = $headers;
```

Comment:
auth and header are always set, and guzzle allow to add request option in the constructor, you can create a dedicated function (or a dedicated class) like 
```php
protected static function buildClient($profile) {
        $opt = array();
        $headers = array();
        $headers['User-Agent'] = 'lizmap-user-agent';

        if (array_key_exists('user', $profile) && array_key_exists('password', $profile)) {
             $opt['auth'] = array();
             array_push($opt['auth'], $profile['user']);
             array_push($opt['auth'], $profile['password']);
        }
        $opt['headers'] = $headers;
        return new new Client($opt);
}
```

and call it in  `uploadToWebDAVStorage `,`deleteFromWebDAVStorage`, ...